src/heartbeat.ts

```diff
@@ -0,0 +1,95 @@
+        this._register(vscode.debug.onDidReceiveDebugSessionCustomEvent(this.updateLastActivitiy, this));
+        this._register(vscode.debug.onDidTerminateDebugSession(this.updateLastActivitiy, this));
+        this._register(vscode.debug.onDidChangeBreakpoints(this.updateLastActivitiy, this));
+        this._register(vscode.extensions.onDidChange(this.updateLastActivitiy, this));
+        this._register(vscode.languages.onDidChangeDiagnostics(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidStartTask(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidStartTaskProcess(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidEndTask(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidEndTaskProcess(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeWorkspaceFolders(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidOpenTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCloseTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidSaveTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidSaveNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidOpenNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCloseNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillCreateFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCreateFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillDeleteFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidDeleteFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillRenameFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidRenameFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeConfiguration(this.updateLastActivitiy, this));
+        this._register(vscode.languages.registerHoverProvider('*', {
+            provideHover: () => {
+                this.updateLastActivitiy();
+                return null;
+            }
+        }));
+    }
+
+    private updateLastActivitiy() {
+        this.lastActivity = new Date().getTime();
+    }
+
+    private async sendHeartBeat(wasClosed?: true) {
+        const suffix = wasClosed ? 'was closed heartbeat' : 'heartbeat';
+        // if (wasClosed) {
+        this.logger.trace('sending ' + suffix);
```

Comment:
we need to send `ide_close_signal` here https://www.notion.so/gitpod/ide_close_signal-e2a0f435fd4745b68567eb4d9afc8fd5

`clientKind` is `vscode`

Reply:
It also implies that telemetry should be disposed after